### PR TITLE
Profile image v3: tracking state of upload

### PR DESCRIPTION
**Pull request by vytis**
## Reason behind this PR

We want to move profile images to use assets v3.

## What's in this PR?

It is first of few PRs for the full implementation. This part adds `UserProfileImageUpdateStatus` that tracks state of profile image upload. At the moment it handles starting the image resize and collecting the results. 
Request strategy for uploading resized images and another one for updating profile will be coming in separate PRs.

P.S. This PR is being merged to intermediate branch (`feature/profile-picture-v3`) where full changes will be collected before merging to `develop`.

### CODE REVIEW

File: Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        }
+        return nil
+    }
+    
+}
+
+extension UserImageAssetUploadStrategy: RequestStrategy {
+    public func nextRequest() -> ZMTransportRequest? {
+        guard let updateStatus = imageUpdateStatus else { return nil }
```
Comment (daehn) [inline Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift]: You might want to pass in a `ClientRegistrationStatus` as well to check if you are logged in etc. before creating a request.

File: Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+extension UserImageAssetUploadStrategy: ZMSingleRequestTranscoder {
+    public func request(for sync: ZMSingleRequestSync!) -> ZMTransportRequest! {
+        if let size = size(for: sync), let image = imageUpdateStatus?.consumeImage(for: size) {
+            return requestFactory.upstreamRequestForAsset(withData: image, shareable: true, retention: .eternal)
+        }
+        return nil
+    }
+    
+    public func didReceive(_ response: ZMTransportResponse!, forSingleRequest sync: ZMSingleRequestSync!) {
```
Comment (daehn) [inline Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift]: I assume the response handling / parsing the asset ID and notifying the status of the finished upload will come in a separate PR?

Comment (vytis) [inline Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift]: Yeah, will come later.

**Comment by mikeger** [comment]
@vytis is this PR up-to-date? Are the comments from @daehn handled?

**Comment by mikeger** [comment]
screen shot 2017-03-14 at 10 39 19

**Comment by daehn** [comment]
@mikeger @vytis as this is an intermediate PR I would suggest merging it so we can track additional progress in the next one.

**Comment by vytis** [comment]
It is currently work-in-progress, so @daehn comments are being addressed at the moment. Just wanted to avoid having a single massive PR

File: Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift (hunk 1 above)
Comment (daehn) [inline Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift]: This comment is still valid and should be fixed at some point (can be in a separate PR to the intermediary branch though).

Comment (vytis) [inline Source/Synchronization/Strategies/UserImageAssetUploadStrategy.swift]: Yes I am currently working on request generating, so will check for authentication status before returning the request

**Comment by vytis** [comment]
So merging it in, more changes coming later today.
